Ensure integrity of middleware-surpassed "promise action" by introduction the check of presence of symbol

jespersandnielsen opened the pull request:
This will fix cannot read properties of undefined for some actions

Error:
```
VM49532:1 Uncaught TypeError: Cannot read properties of undefined (reading 'resolved')
    at eval (eval at isTriggerAction (vendors-node_modules…cff275.chunk.js:1:1), <anonymous>:1:29)
    at isTriggerAction (index.ts:83:1)
    at Object.dispatch (index.ts:258:1)
    at dispatch (<anonymous>:6:7384)
    at Object.initialize (redux.js:546:1)
    at Form._this.initIfNeeded (createReduxForm.js:141:1)
    at Form.UNSAFE_componentWillMount (createReduxForm.js:512:1)
    at callComponentWillMount (react-dom.development.js:12788:1)
    at mountClassInstance (react-dom.development.js:12878:1)
    at updateClassComponent (react-dom.development.js:17426:1)
```

teneko commented:
Mhh interesting case. Assuming that others may also compose an action with meta.promiseActions, I accidentally shortcutted the validation logic by just checking the existence of one componenet, here namely 'resolved'. That definitively needs to be fixed. Thanks for pointing out. :)

teneko reviewed (changes requested):
This needs to be expanded by either two ideas. Either we annotate each created promose action from us with a global constant symbol, so we only check for that or we check for the hole form of the promise action, which includes existence of promiseActions.resolved, promiseActions.rejected and being themselves actions. The symbol approach seems the most easiest. The only disadvantage is, that someome who is mimicking our promise action must now be aware of that symbol which is implictly a breaking change. What do you think?

jespersandnielsen commented:
@teneko Hmm. I agree the symbol solution sounds like the best solution, taking maintainability into consideration, but yes it would include a breaking change for people mimicking the library. Possibly we should release the symbol change under a new major version?

jespersandnielsen commented:
I also had some issues with the redux middleware not picking up the merged action in the middleware, however I hadn't found a solution for this issue yet

jespersandnielsen commented:
@teneko I have moved module resolution to global map, such that it doesn't require mutable actions in middleware, since this is not working for me, and we get lookups in O(1) time. furthermore I have introduced a global symbol to determine if the action is a redux saga promise action

teneko commented on src/index.ts:
```diff
@@ -157,6 +164,7 @@ function wrapTriggerAction<V, P, T extends string, TA extends PayloadActionCreat
 
   const updatedTrigger = <TriggerActionCreator<V, P, T, SagaPromiseMeta<V, T>, TA>>createAction(type, (...args: any[]) => merge(triggerAction.apply(null, args), {
     meta: <SagaPromiseMeta<V, T>>{
+      actionSymbol: promiseActionSymbol,
```
One idea of symbols is to use a symbol as property key to have a unique property key that only we, the authors of the package can associate to and for example check for uniqueness of an object. Instead of "actionSymbol" = promiseActionSymbol, you would rather use [promiseActionSymbol] = true. 🙂

teneko replied:
By introducing another clearname property we face the same exact limitations as with `meta.promsieActions` or `meta.promiseActions.resolved`.

jespersandnielsen replied:
@teneko fixed

teneko commented on src/index.ts:
```diff
@@ -79,26 +80,32 @@ export type SagaPromiseActionCreator<V, P, T extends string, TA extends PayloadA
+
 function isTriggerAction(action: SagaPromiseAction<any, any, any>) {
-  return action?.meta?.promiseActions.resolved != null;
+  return action?.meta?.actionSymbol === promiseActionSymbol;
```
Here you would have to check the symbol to exist as property key and optionally for truthyness, since we assigned `true`.

jespersandnielsen replied:
@teneko fixed

teneko commented on src/index.ts:
```diff
@@ -2,32 +2,33 @@ import { Dispatch, Middleware, MiddlewareAPI } from "redux";
 import { ArgumentError } from "./ArgumentError";
 import { ConfigurationError } from "./ConfigurationError";
 
+const promiseActionSymbol = Symbol("redux-saga-promise-action");
```
It is mandatory to export it to make it globally available, so we allow others still to mimick our action structure.

jespersandnielsen replied:
@teneko fixed

teneko commented on src/index.ts:
```diff
@@ -79,26 +80,32 @@ export type SagaPromiseActionCreator<V, P, T extends string, TA extends PayloadA
 
 export type SagaPromisePreparedActionCreator<V, T extends string, TA extends PrepareAction<any>> = SagaPromiseActionCreator<V, ReturnType<TA>["payload"], T, _ActionCreatorWithPreparedPayload<TA, T>>;
 
+let promiseResolutionMap: PromiseResolutionMap<any> | undefined;
```
This is an absolute no-no. We cannot shift the `promiseResolution` to the scope of the module nor we ever should let multiple actions reassign a module-scoped variable, here namely `promiseResolution`.

teneko replied:
`promiseResolution`is absolutely tight to one promsie action, imagine it as 1:1 relationship.

jespersandnielsen replied:
@teneko what do you suggest?

The issue here is that somehow the promiseResolution is not available on the trigger action.

teneko replied:
This sounds like the initially created promise action has not surpassed the middleware? :o
Could you create an absolute minimal reproduction of your actual problem?

teneko replied:
We come to a point I cannot tell for sure where the actual problem is lying. Does it help you to explain the problem, by inspecting the logic implemented here: https://github.com/teneko/redux-saga-promise/blob/d148fb427b3de17f1dcb9edaf6f56214bba4489d/src/index.ts#L152 ?

jespersandnielsen replied:
@teneko I have created a codepen here with the issue https://codepen.io/jespersandnielsen/pen/jOoNpZx
It throws this error, as you can see in the console
```
core.mjs:2 Error: redux-saga-promise: rejectPromiseAction: Unable to execute--it seems that promiseMiddleware has not been not included before SagaMiddleware
    at b (redux-saga-promise.mjs:6:2164)
    at C (redux-saga-promise.mjs:6:2675)
    at C.next (<anonymous>)
    at d (core.mjs:9:1321)
    at G (core.mjs:9:1150)
    at Hn (core.mjs:6:5104)
    at m (core.mjs:9:1564)
    at E (core.mjs:9:1994)
    at d (core.mjs:9:1382)
    at Object.k [as cont] (core.mjs:9:1874)
le @ io-22ea0cf9.js:111
Show 1 more frame
Show less
core.mjs:2 The above error occurred in task C
    created by handler
    created by takeLatest(promiseAction, handler)
    created by watcher
Tasks cancelled due to error:
takeLatest(promiseAction, handler)
```
because the actions doesn't include promiseResolution
```
{
  "type": "promiseAction",
  "meta": {
    "promiseActions": {
       "rejected": fn(),
       "resolved": fn(),
     }
  }
}
```

jespersandnielsen replied:
@teneko have you had the time to look at my codepen yet?

teneko replied:
I am going to inspect this today ! Sorry for the delay.

jespersandnielsen replied:
No worries

teneko replied:
I need to shift this to saturday. Lot of work.

teneko replied:
Thanks for your input and work so far ! This needs definitvely to be investigated and solved !

jespersandnielsen replied:
Sure no worries

teneko replied:
Oh damn, I think you misunderstood how actions work. :(

Similiar to the thing you just assumed reassigning to `let promsieResolutionMap`is fine you unfortunately thought that you can also use a **new** action, by using the action creator in a saga handler, this is not how actions are handled in a saga handler.

As soon as you dispatch a **new** action, by doing `promiseAction()` then this action, and by this I mean the instance you just created is pass-throughed the redux saga middleware and therefore redux sagas and therefore also passed in your defined handler method.

So please look at the following code:

```typescript
import createSagaMiddleware from "redux-saga";
import { call, takeLatest } from "redux-saga/effects";
import { configureStore, createReducer } from "@reduxjs/toolkit";
import {
  promiseMiddleware,
  promiseActionFactory,
  resolvePromiseAction,
} from "./index";

// Create promise action **creator**
const promiseAction = promiseActionFactory<string>().create("promiseAction");

function* handleTriggerAction(action: any) {
  try {
    yield call(resolvePromiseAction, action, "works");
    // or
    // yield call(promiseAction.sagas.resolve, action, "works");
  } catch (e) {
    yield call(promiseAction.sagas.reject, action, e.message);
  }
}

function handleResolvedAction(action: typeof promiseAction.types.resolvedAction) {
  console.log(`${action.payload} (hello from saga)`);
}

function* watcher() {
  yield takeLatest(promiseAction, handleTriggerAction);
  yield takeLatest(promiseAction.resolved, handleResolvedAction);
}

const reducer = createReducer<string>("<store-yet-uninitialized>", (builder) => {
  builder.addCase(promiseAction.trigger, (state) => state);
  builder.addCase(promiseAction.resolved, (_unused_state, action) => `${action.payload} (hello from store)`);
  builder.addCase(promiseAction.rejected, (state) => state);
});

const sagaMiddleware = createSagaMiddleware();

// createStore is deprecated, so use:
const store = configureStore({
  reducer,
  middleware: [promiseMiddleware, sagaMiddleware],
});
// or
// const store = legacy_createStore(reducer, middleware);

sagaMiddleware.run(watcher);

store.subscribe(() => {
  console.log(store.getState());
});

// promiseAction() leads to a NEW instance,
// which first gets transformed to a fully qualified promise action by OUR middleware
// and then THIS NEW INSTANCE is passed through the redux saga :)
store.dispatch(promiseAction()).then((result) => console.log(`${result} (hello from promise)`));
```

You simply can paste this code to "/src/testCase.ts"

and run it by:

```
$ npm install -g ts-node
$ ts-node /src/testCase.ts
```

You will see this output:

```txt
<store-yet-uninitialized>
works (hello from store)
works (hello from saga)
works (hello from promise)
```

teneko commented:
If you still want to see your effort regarding the symbol approach in the library, then please open a new pull request with just the changes regarding the symbols approach. Please do not change any TypeScript types. :)

teneko commented:
Please do not change or update the package.json nor package-lock.json. Dependency updates should always be lye on the side of the developers. 🙂

teneko commented:
Version 3.0.1 of @teneko/redux-saga-promise has been released. Please feel free to take a look into it. The use of resolvePromiseAction, \<any-promise-action-creator\>.sagas.resolve and the trigger and reject pendants should now consistenly throw the desired error if a non-middleware-surpassed promise-action is used.
